Convert "Create a DevTools extension" to sample-based (#3616)
MicrosoftDocs/edge-developer: microsoft-edge/extensions/samples/custom-devtools-tool.md

---
title: "Sample: Custom DevTools tool"
description: Microsoft Edge extension that adds a Custom tool in DevTools.  Sends messages between the inspected webpage and DevTools.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: article
ms.service: microsoft-edge
ms.subservice: extensions
ms.date: 04/20/2026
---
# Sample: Custom DevTools tool
<!-- https://learn.microsoft.com/microsoft-edge/extensions/samples/custom-devtools-tool -->
<!-- sibling 2 of 3 -->

<!-- todo: move this new file from /samples/ to /developer-guide/ -->

> The sample "Custom DevTools tool" is a Microsoft Edge extension that adds a **Custom** tool in Microsoft Edge DevTools, including a tab in the **Activity Bar**, and a panel below the tab.  The tool name and its tab in DevTools can be something other than **Custom**.

Reviewer: ```suggestion
The sample "Custom DevTools tool" is a Microsoft Edge extension that adds a **Custom** tool in Microsoft Edge DevTools, which has its own tab in the **Activity Bar**.
```

A tab and its panel always go together 1 to 1. A tab can't exist without its panel. So I think we can remove "and a panel below the tab".

I'm also suggesting to remove the last sentence about the custom name. I don't think it's useful in this intro.

* The **Custom** DevTools tool calls the DevTools API to display memory information.

* The webpage under inspection, and the **Custom** DevTools tool, send messages to each other.

Download, install, use, and modify the sample, as follows.

See also:
* [Add a custom tool in Microsoft Edge DevTools](../developer-guide/devtools-extension.md)<!-- sibling 1 of 3 -->
* [Sample: Custom DevTools tool](./custom-devtools-tool.md)<!-- sibling 2 of 3 --> - the present article.

> * [Code for Custom DevTools tool](./custom-devtools-tool-code.md)<!-- sibling 3 of 3 -->

Reviewer: This list of links might not be needed anymore if we remove the first article.
I would simplify to:

```suggestion
Use this article to download, install, and run the sample. To learn more about the code in this sample, see [Code for Custom DevTools tool](./custom-devtools-tool-code.md).
```

<!-- ------------------------------ -->
#### Display memory information

The **Custom** tool in DevTools calls the DevTools API (`chrome.system.memory.getInfo((data)`) to retrieve and display memory information.

<!-- ------------------------------ -->
#### Send message from DevTools to inspected page

The **Custom** tool in DevTools has a button that sends a message to the inspected webpage.  The inspected webpage then displays a JavaScript alert dialog:

![Alert dialog](./custom-devtools-tool-images/alert.png)

<!-- ------------------------------ -->
#### Send message from inspected page to DevTools

Clicking around within the inspected webpage sends a message to DevTools.  The **Custom** tool uses a content script to send a message (the clicked coordinates) from the inspected webpage (any webpage) to the **Custom** DevTools tool and the **Console** tool in DevTools.

The **Custom** tool displays the mouse-click coordinates from the message:

![Coordinates in the Custom tool](./custom-devtools-tool-images/custom-tool-coordinates.png)

The **Console** tool also displays the mouse-click coordinates from the message:

> ![Coordinates in the Console tool](./custom-devtools-tool-images/console-tool-coordinates.png)

Reviewer: The user has not installed the sample yet. So, describing its features in so much details, so early, feels wrong. Plus, we already do this in steps 4, 5, and 6.

Can we turn this into a shorter description?

```suggestion
The **Custom** tool displays memory information and sends messages between the inspected webpage and the panel in DevTools.
```

The first thing users will want to do when reading this article is install the extension and play with it. So I would keep the intro sections as short as possible to avoid delaying them from doing that.

1. In the **Activity Bar** of DevTools, click the **Custom** (![The Custom tool icon](./custom-devtools-tool-images/custom-icon.png)) tool's tab.

   The **Custom** tool tab and panel are displayed:

   ![Custom tab](./custom-devtools-tool-images/custom-tab-activity-bar.png)

   If the **Custom** (![The Custom tool icon](./custom-devtools-tool-images/custom-icon.png)) tool's tab isn't visible, do any of the following:

   * Click the **More tools** (![More tools icon](./custom-devtools-tool-images/more-tools-icon.png)) button, and then select ![Custom icon](./custom-devtools-tool-images/custom-icon.png) **Custom**.

   * Make DevTools wider, and then click the **Custom** (![Custom icon](./custom-devtools-tool-images/custom-icon.png)) tool's tab.

   The custom DevTools page has several sections:

   * Memory display information.

   * A button to send a message from DevTools to the inspected webpage, to make the page display a JavaScript `alert` dialog.

   * A **Coordinates** display area, to send a message from the inspected webpage to the DevTools **Console** and **Custom** tools.

<!-- ====================================================================== -->

> ## Step 4: View memory information from a DevTools API call

Reviewer: ```suggestion
## Step 4: View memory information by using an extension API call
```

   ![/devtools-extension/ folder in Visual Studio Code](./custom-devtools-tool-images/panel-html.png)

1. Add "My" to the `<h2>` heading; change from:

   ```html
   <h2>Custom DevTools Tool</h2>
   ```

   to:  

   ```html
   <h2>My Custom DevTools Tool</h2>
   ```

1. Save `panel.html`.

1. In the **Explorer** pane, click `manifest.json`.

   `manifest.json` is opened for editing.

1. Add "My" to the `name`; change from:

   ```json
   "name": "Custom DevTools Tool",
   ```

   to:  

   ```json
   "name": "My Custom DevTools Tool",
   ```

1. Save `manifest.json`.

<!-- ====================================================================== -->
## Step 8: Reload the modified Custom tool

1. In Microsoft Edge, select **Settings and more** (![The Settings and more icon](./custom-devtools-tool-images/settings-and-more-icon.png)), hover over **Extensions**, and then select **Manage extensions**.

   The **Extensions** tab and page opens (`edge://extensions`).

1. In the **Custom DevTools Tool** card, click the **Reload** link.

   The card now shows **My Custom DevTools Tool** as the name of the extension:

   ![Reload](./custom-devtools-tool-images/reload-modified-sample.png)

>    An **Errors** link might appear in the card, displaying "Uncaught Error: Extension context invalidated", because the sample was modified while still in use.  If you want to avoid this error, you can click **Remove** in the **Custom DevTools Tool** card, and then click **Load unpacked**.

Reviewer: I think we need to remove this para. The developer experience here is: 1. load unpack 2. make a change 3. click reload.
Asking developers to remove, then load unpack again isn't great. If errors are preventing them from using the simpler developer experience, then they can submit them to us via the feedback dialog, and we fix that.
